core/shared/src/main/scala-2/cats/effect/cpsinternal/AsyncAwaitDsl.scala

```diff
@@ -99,7 +99,7 @@ object AsyncAwaitDsl {
           }
           ${F}.flatten {
             _root_.cats.effect.std.Dispatcher.sequential[$effect].use { dispatcher =>
-              ${F}.async_[$name#FF[AnyRef]](cb => new $name(dispatcher, cb).start())
+              ${F}.async[$name#FF[AnyRef]](cb => ${F}.delay { new $name(dispatcher, cb).start(); Some(${F}.unit) })
```

Comment:
I'm wary of the `F.unit` call here : I think what should happen instead is some mutation to the `StateMachine` that could/would be inspected in any subsequent step in the state machine to short circuit the computation 

it'd look like : 

```
${F}.delay {
  val stateMachine = $name(dispatcher, cb) 
  stateMachine.start()
  Some(${F}.delay(stateMachine.cancel()))
} 
```

Then in `AsyncStateMachine`, you'd have something like 

``` scala
private val isCancelled = new AtomicBoolean(false)

def cancel() : Unit = isCancelled.set(true) 

def onComplete(f: F[AnyRef]) : Unit = {
  // short-circuits the next call to dispatcher 
  if (isCancelled.get) {
    // surfaces the cancellation to the `async` invocation. 
    this(Left(F.canceled.asInstanceOf[F[AnyRef]]))
  } else dispatcher.unsafeRunAndForget(...)
} 
```

Reply:
@djspiewak, mind giving a sanity to what I said above ? Also maybe an `AtomicBoolean` is overkill and a simple `@volatile var boolean` might do the trick.

Reply:
I think this is exactly correct. What's being revealed here is the fact that `AsyncStateMachine` at present has no support for cancelation, but the old properties of `async_` caused that reality to be ignored, effectively leaking fibers (and potentially even creating thread starvation issues) when `async`/`await` blocks were canceled.

Also yes, a volatile should be mostly sufficient in this case. You're really just looking for a flag to make it stop. The one trick though is you'll need to make sure the finalizer itself asynchronously blocks until the state machine has been fully interrupted, which obviously might take a bit.